r/R/feather.R

```diff
@@ -0,0 +1,160 @@
+}
+
+#' Write data in the feather format
+#'
+#' @param data frame or arrow::RecordBatch
+#' @param stream A file path or an arrow::io::OutputStream
+#'
+#' @export
+write_feather <- function(data, stream) {
+  UseMethod("write_feather", data)
+}
+
+#' @export
+write_feather.default <- function(data, stream) {
+  stop("unsupported")
+}
+
+#' @export
+write_feather.data.frame <- function(data, stream) {
+  write_feather(record_batch(data), stream)
+}
+
+#' @method write_feather arrow::RecordBatch
+#' @export
+`write_feather.arrow::RecordBatch` <- function(data, stream) {
+  write_feather_RecordBatch(data, stream)
+}
+
+#' @rdname write_feather
+#' @export
+write_feather_RecordBatch <- function(data, stream) {
+  UseMethod("write_feather_RecordBatch", stream)
+}
+
+#' @export
+#' @method write_feather_RecordBatch default
+`write_feather_RecordBatch.default` <- function(data, stream) {
+  stop("unsupported")
+}
+
+#' @export
+#' @method write_feather_RecordBatch character
+`write_feather_RecordBatch.character` <- function(data, stream) {
+  `write_feather_RecordBatch.fs_path`(data, fs::path_abs(stream))
+}
+
+#' @export
+#' @method write_feather_RecordBatch fs_path
+`write_feather_RecordBatch.fs_path` <- function(data, stream) {
+  file_stream <- close_on_exit(file_output_stream(stream))
+  `write_feather_RecordBatch.arrow::io::OutputStream`(data, file_stream)
+}
+
+#' @export
+#' @method write_feather_RecordBatch arrow::io::OutputStream
+`write_feather_RecordBatch.arrow::io::OutputStream` <- function(data, stream) {
+  ipc___TableWriter__RecordBatch__WriteFeather(feather_table_writer(stream), data)
+}
+
+#' A arrow::ipc::feather::TableReader to read from a file
+#'
+#' @param file A file path, arrow::io::RandomAccessFile
+#' @param mmap Is the file memory mapped (applicable to the character and fs_path methods)
+#' @param ... extra parameters
+#'
+#' @export
+feather_table_reader <- function(file, ...){
+  UseMethod("feather_table_reader")
+}
+
+#' @export
+feather_table_reader.default <- function(file, ...) {
+  stop("unsupported")
+}
+
+#' @export
+feather_table_reader.character <- function(file, mmap = TRUE, ...) {
+  feather_table_reader(fs::path_abs(file), mmap = mmap, ...)
+}
+
+#' @export
+feather_table_reader.fs_path <- function(file, mmap = TRUE, ...) {
+  stream <- if(isTRUE(mmap)) mmap_open(file, ...) else file_open(file, ...)
+  feather_table_reader(stream)
+}
+
+#' @export
+`feather_table_reader.arrow::io::RandomAccessFile` <- function(file, ...){
+  unique_ptr(`arrow::ipc::feather::TableReader`, ipc___feather___TableReader__Open(file))
+}
+
+#' @export
+`feather_table_reader.arrow::ipc::feather::TableReader` <- function(file){
+  file
+}
+
+#' Read a feather file
+#'
+#' @param file a arrow::ipc::feather::TableReader or whatever the [feather_table_reader()] function can handle
+#' @param ... additional parameters
+#'
+#' @return an arrow::Table
+#'
+#' @export
+read_feather <- function(file, ...){
```

Comment:
It occurs to me that the semantics of this function are different from `feather::read_feather` (it returns an ``arrow::Table`` instead of `data.frame`), if we care

Reply:
One of the goals of `arrow` is to make Array structures palpable in R, which feather didn't do. But yeah I guess we need both, i.e.
 - a function that returns an `arrow::Table`, not sure how to call this one. 
 - a function, perhaps `read_feather` if this need to look like `feather::read_feather` that returns a tibble by calling as_tibble on the result of the first one. 

Right now, it makes sense to go straight to tibble, but soon enough this (or other) packages will have more tooling to manipulate and compute directly on arrow memory.

Reply:
You could add an `as_tibble = TRUE` argument to `read_feather` so this can be toggled on and off